java/src/jmri/util/PipeListener.java

```diff
@@ -27,9 +27,11 @@ public void run() {
             while (true) {
                 try {
                     c[0] = (char) pr.read();
-                    ta.append(new String(c));   // odd way to do this, but only
-                    // way I could think of with only one
-                    // new object created
+                    jmri.util.ThreadingUtil.runOnGUI(() -> {
+                        ta.append(new String(c));   // odd way to do this, but only
+                        // way I could think of with only one
+                        // new object created
+                    });
```

Comment:
No need to change, but this method would be much more efficient if the characters are read to a buffer instead of one character at a time.

https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/io/PipedReader.html#read(char%5B%5D,int,int)

Reply:
It's been a long time since I worked on this (2005!), so perhaps JVM behavior has changed, but the original problem was that the `read(char[] cbuf, int off, int len)` call would not return immediately with the data it had if it had less than `len` bytes and the pipe was still open.  In other words, the last few characters of a message got stuck waiting for more content to push the buffer. AFAIK, the behavior for this case isn't documented anywhere. I'll try it out and see what's happening with J11.

Reply:
@danielb987 Thanks for the suggestion.  I've pushed a new algorithm for this. It seems to be working on Mac.  Do you have a way to try it on Windows?  I think the Windows tests run in CI will check for the not-partial-read case, but would feel better if it was actually tried before merge. Thanks again.

Reply:
Yes, I can test it on my Windows 8.1 computer. I have downloaded your branch and compiles it now. How do I test this? Where in JMRI is this used?

Reply:
1) Open the Script Entry and Script Output windows
2) Enter something, e.g. "print (123)" in the entry window and hit Execute.
3) Do you see the complete output in the Output window?  
Try with a couple other inputs with varying lengths.

Thank you!

Reply:
Something is not working right.

PanelPro version 5.5.3plus+daniel+2023-08-30T13:37:30Z+R2f736fa starts under Java 13.0.2 on Windows 8.1 amd64 v6.3

See [the file](https://github.com/JMRI/JMRI/pull/12331/files) in PR #12331.

When I have multiple lines with long content, like 50 lines of:
`print "This is a long string to test script output"`
I get a blocking on `ThreadingUtil.runOnGUI()`.

The `System.out.format` at line 47 prints but not the `System.out.format` at line 50 in my PR. So the problem doesn't seem to be with `PipeListener` but with GUI thread being blocked.

Reply:
> Thanks again, and sorry for the trouble.

No problem! Happy to help!

With `runOnGUIEventually`, everything works just fine with your PR. All my tests worked as expected. And I tested it hard.  One of my tests printed a 270 character long string and it had 190 such print statements. Without any problem.

I need to take a break but will be back in a couple of hours and can do more testing then.

Reply:
I've updated the branch with `RunOnGuiEventually`.   It works properly on my end.  When you get the chance (no hurry), I'd appreciate it if you could try that to confirm.

Reply:
I have run all my tests and it works fine.

I have also verified that the old code, pre this PR, also worked fine. But this PR is much faster when it's a huge output.

One question: Is there an easy way to add a test to `jmri.util.PipeListenerTest` that verifies that PipeListener doesn't block the GUI thread? The problem I discovered seems to only happens when there is a large output. If someone later changes `runOnGuiEventually` to `runOnGui`, it might cause bugs that's difficult to track down if you don't understand the cause.

Reply:
I don't know how to write a test like that.  What I'll do is add a comment on the line where the next person will see it.

Reply:
@bobjacobsen 
> I don't know how to write a test like that. What I'll do is add a comment on the line where the next person will see it.

Thanks.

See PR #12332. It works with `runOnGuiEventually` but fails with `runOnGui`.